### PR TITLE
[mtoh] Refactor settings / defaultRenderGlobals storage

### PR DESCRIPTION
**Pull request by marsupial**
There are a few issues with the current render-settings / defaultRenderGlobals, that make it difficult for a delegate to merge _USDRenderSettings_ and _HdDelegate::SetRendererSetting_.  The larger goal of editing a USDRenderSettings and applying a delegates schema aren't dealt with here, as the PR seems large enough just to address the current issues:

- Largest issue is that if a delegate provides a USD-namespaced setting, **mtoh** will error out when trying to set it.
- If any two delegates have the same setting (because they can't namespace) then they will stomp one another.
- User state isn't honored when not saved into a scene, this is problematic for jumping between files as the user may have the delegate up in a specific way optimized for the task.
- Updates to a single setting incur a large cost of synchronizing all settings across all delegates.
- Excessive copies of _MtohRenderGlobals_, when one is more than sufficient.
- Changes from editing values from via Attribute Editor do not make it to the delegate.
- Need proper layout/grouping in the Attribute Editor as well.

mtoh-ae-before
mtoh-hydra-group
mtoh-groups-expaned

### CODE REVIEW

**Review by pmolodo — CHANGES_REQUESTED** [review]
Mostly looks good - I definitely like the overall goal here, so glad to have this PR, and thank you!

Just a few small typos / comments / questions...

**Comment by kxl-adsk** [comment]
@marsupial Have a look as well at the build errors with MacOS - https://github.com/Autodesk/maya-usd/pull/758/checks. You can get build logs by downloading the artifact associated with the check.

File: lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp
Comment (marsupial) [reply lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:193]: No problem changing them to whatever (maybe @kxl-adsk can weigh in too), but I'd prefer separate lines at least:
```
if (!status) {
  return false;
}
```

* Matches the rest of the repo better
* Single line style doesn't work well with stepping through in visual-debuggers.

File: lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp
Comment (marsupial) [reply lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:424]: I'll see if the template makes this guy easier to read, but generally curious if dropping the GfVec4 variant all-together is better ?
There is one Storm selection setting that does rgba, but the rest of Hydra & Shading for USD seems to be discouraging any use of rgba.

@mattyjams has an opinion on this ?

File: lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp
Comment (marsupial) [reply lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:536]: Yes it is being used internally to adopt a Houdini-esque mechanism where you can target a RenderSetting prim, which then sets a lot of parameters at once.

File: lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp
Comment (pmolodo) [reply lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:424]: Hmm - yeah, if we can get away with dropping the vec4, that would be best. Hadn't looked to see if that was an option...

File: lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp
Comment (pmolodo) [reply lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:536]: Hmm... gotcha, I was hoping we could drop the c-style array usage.  Well - as a compromise, what do you think about making this a private function, and then make public wrapper functions - one of which takes an optional attr name (ie, 0 or 1 attrs), and the other takes a TfTokenVector? (Or perhaps some other type, if it's more appropriate to your use case?)

File: lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp
Comment (pmolodo) [reply lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:193]: Well, I may have gotten ahead of myself, given that we're still having [active debate](https://github.com/Autodesk/maya-usd/pull/690) , so I'll leave this one to your discretion for now.  Once we come to a consensus, we can fix everything at once when we do our planned massive-style-update.

File: lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp
Comment (marsupial) [reply lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:45]: I'll see if the template makes this guy easier to read, but generally curious if dropping the GfVec4 variant all-together is better ?
There is one Storm selection setting that does rgba, but the rest of Hydra & Shading for USD seems to be discouraging any use of rgba.

@mattyjams has an opinion on this ?

File: lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp
Comment (marsupial) [reply lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:566]: Think it's accurate, as `mtohRenderOverride_AddAttribute` is in **_renderOverride_PreAmble**, but can just dump the text that failed (if this error should trigger, something has really gone wrong anyway)

File: lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp
Comment (mattyjams) [reply lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:424]: Not a particularly well informed one I'm afraid, as I'm not that familiar with the mayaToHydra code, and not really in the loop on the Hydra-side work in this area.

But there has been a trend recently to separate out the handling of color and opacity, for example the change here:
https://github.com/PixarAnimationStudios/USD/commit/332db31e5c0c631e404d5f20c17e8e0d6b30a630

That doesn't extend everywhere yet as you pointed out though, since `UsdImagingGLEngine` and `HdxTaskController` at least so far still expect to be given the selection color as a `GfVec4f`.

So yeah, not sure if that helps at all. :/

File: lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp
Comment (mattyjams) [reply lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:45]: Not a particularly well informed one I'm afraid, as I'm not that familiar with the mayaToHydra code, and not really in the loop on the Hydra-side work in this area.

But there has been a trend recently to separate out the handling of color and opacity, for example the change here:
https://github.com/PixarAnimationStudios/USD/commit/332db31e5c0c631e404d5f20c17e8e0d6b30a630

That doesn't extend everywhere yet as you pointed out though, since `UsdImagingGLEngine` and `HdxTaskController` at least so far still expect to be given the selection color as a `GfVec4f`.

So yeah, not sure if that helps at all. :/

File: lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp
Comment (marsupial) [reply lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:536]: TfTokenVector works for me (c-style probably exists as I thought the heap allocation would be frowned on).

Comment (pmolodo) [reply lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:536]: Well, I was guessing that most of the time you already have a vector or something holding your tokens, so it wouldn't be an extra allocation.  But even if you do have to, I don't think render globals change frequently enough for it to be a big deal... I'd rather have the memory safety.

File: lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp
Comment (marsupial) [reply lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:45]: At least clarifies that supporting GfVec4 needs to stay for a while, and probably won't reduce much complexity having to support version before anyways...

So yeah, thanks!

**Comment by marsupial** [comment]
Resolved all the 'simple' stuff as I went through them, but let the longer comments open

File: lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp
Comment (kxl-adsk) [reply lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:194]: @marsupial, @pmolodo - I would recommend matching the rest of the style in mtoh module for now. We don't have to block the PR by the discussion we are having on the official style guidelines for a single line if statements and how to enforce it.

Comment (marsupial) [reply lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:194]: I wound up doing the above (mostly for the step-debug issues.
If style is a separate topic, then hopefully when that's resolved a clang-format on the repo mitigates any differences.

**Comment by kxl-adsk** [comment]
@marsupial I'm trying to get a sense of where we are with this PR? Your last comment mentions that there are still unresolved comments. As well, did you fix mac build errors mentioned here - https://github.com/Autodesk/maya-usd/pull/758#issuecomment-694244521

**Comment by marsupial** [comment]
They've all been addressed as far as I'm aware.

**Comment by kxl-adsk** [comment]
@pmolodo Anything is still missing before you can approve this PR?

File: lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp
Comment (pmolodo) [inline lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:154]: Another braceless-if

Now that we've decided how to replace these, there are a number of them still in the code - I found that if you just search for this regexp:

`if *\([^\n{]+\n`

...it should find them (as well as a few multi-line-if red-herrings, but not many).

File: lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp
Comment (pmolodo) [reply lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:194]: Commented on line 154 where there's another example, but I'll repeat here for clarity:

Now that we've decided how to replace these, there are a number of them still in the code - I found that if you just search for this regexp:

if *\([^\n{]+\n

...it should find them (as well as a few multi-line-if red-herrings, but not many).

File: lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp
Comment (pmolodo) [inline lib/mayaUsd/render/mayaToHydra/renderGlobals.cpp:361]: Missed this rename to `postCreate`...

**Review by pmolodo — CHANGES_REQUESTED** [review]
Basically there - all the more significant changes were made, just a typo, missed rename, and a few more instances of braceless-ifs left.

**Review by pmolodo — APPROVED** [review]
Looks good!

**Pull request merged**
